### ISSUE TITLE
Easy Map Generator for VWP Projects

### ISSUE THREAD

**Issue by jdkleiner** [HARPgroup/hydro-tools#369]
Easiest Map Generation (Rev. 5/4/23):

Everything needed to generate a "fig.location_map_.png" is housed in this file: 
- Original: https://github.com/HARPgroup/hydro-tools/blob/master/GIS_functions/EasyMapGen.R
  - Propose to develop in vahydro since we don't want to develop in hydrotools as that is to be reserved for accessing rest data and limited hydro analysis.  See here: https://github.com/HARPgroup/vahydro/blob/master/R/mapping/EasyMapGen.R
- This is the most updated version of the mapping script
- This file currenlty contains the updated function definitions for `mapgen()` and `model_geoprocessor()` as updates were needed during development and I didn't want to update those functions located in `hydro-tools/GIS_functions/` just yet in case you needed them
- To generate a map, a user only needs to supply a point location to base the map on (i.e. the intake location) and a dataframe of all the additional points they want mapped (See EasyMapGen.R)

Result: 
![fig location_map_test](https://user-images.githubusercontent.com/29379385/236312194-5f29268f-5850-478c-a92f-e6b38bdab999.png)

**Comment by jdkleiner** [comment]
Stashing pre-5/4/23 items below: 

### Scripts:
- runner: https://github.com/HARPgroup/hydro-tools/blob/master/examples/map_runner.R
   - calls https://github.com/HARPgroup/hydro-tools/blob/master/GIS_functions/mapgen.R
      - Uses new `nhdplusTools` mapping functions, also applies the new `model_geoprocessor()` function: https://github.com/HARPgroup/hydro-tools/blob/master/GIS_functions/model_geoprocessor.R

![example_map](https://user-images.githubusercontent.com/29379385/198049239-381a484e-a7ca-4b5a-bd57-e27f4cc71e94.png)


### Example:
```r 
source("https://raw.githubusercontent.com/HARPgroup/hydro-tools/master/GIS_functions/mapgen.R")

# User Inputs: 
export_path <- "C:/Users/nrf46657/Desktop/VWP Modeling/Magnolia Green"
filename <- paste0("MagnoliaGreen_nhdplus_map.png")

# specify start point (typically intake location), map buffer is based on this point
start_point <- data.frame(lat = 37.415128974720155, lon = -77.7271085761693, label = "Intake")

# specify additional points to plot
points <- data.frame(lat=double(),lon=double(),label=character())
points <- rbind(points,data.frame(lat = 37.40402068694781, lon = -77.74288001411728, label = "12MG Pond"))
points <- rbind(points,data.frame(lat = 37.402823839033694, lon = -77.74452198900777, label = "7MG Pond"))

# specify usgs gage to plot
gageid <- "02036500"

# specify which rsegs to plot
segswhere <- "hydrocode LIKE 'vahydrosw_wshed_J%'"

# generate map gg object (simple example, using defaults)
# map_gg <- mapgen()

# generate map gg object (simple example, overriding defaults)
map_gg <- mapgen(start_point=start_point,
                 points=points,
                 gageid=gageid,
                 segswhere=segswhere)

# output map as png file
png(file=paste(export_path,filename,sep="/"), width=1500, height=1500)
map_gg
dev.off()
```

### Magnolia Green Example:
![DEMO_MAP_2 0](https://user-images.githubusercontent.com/29379385/219785287-bc0b038b-a6c9-4a43-b6d6-1453038f4da9.png)

### Salem WTP Example:
```r
map_gg <- mapgen(start_point=data.frame(lat = 37.286388888900, lon = -80.075833333300, label = "Intake"))
```
![TEST_map](https://user-images.githubusercontent.com/29379385/219794193-7ee53484-0c96-4a1e-b735-e8ae57ae4e47.png)

**Comment by rburghol** [comment]
Hey - the new model_geoprocessor function looks very good.  I propose the following: (Not really proposing that *you* do it, just proposing that *we* do it some time :) )
- remove the SQL segs_where portion from the function (function should either retrieve data or display the map)
- use the function `fn_extract_basin()` to grab those that we want.  Example: see below (note: dataframe funkiness requires that you have 2 columns, riverseg and *anything else* for one of the syntaxes to work)

```
# Three ways to get the list of segments:
# 1. Use the dataframe returned from om_vahydro_metric_grid()
#   - can be useful as it will ONLY get segments that have been modeled for the scenario
#   - can be problematic cause it ONLY gets segments that have been modeled :)
# GET RIVERSEG l90_Qout DATA
conf <- data.frame( model_version ='vahydro-1.0',  runid = 'runid_600', metric='l90_Qout', runlabel='l90_prop')
om_data <- om_vahydro_metric_grid(
  metric = False, runids = conf, 
  base_url = paste(site,'entity-model-prop-level-export',sep="/"),
  ds = ds
)
seglist <- data.frame(riverseg = wshed_data$riverseg, propname = wshed_data$propname)
# 2. Use ds$get() call
#    - this is a touch slower, but very concise
seglist <- ds$get('dh_feature', config=list(ftype='vahydro',bundle='watershed'))
# 3. There is also a views public query available which should be super fast

# Then, extract the basin using fn_extract_basin()
app_segs <- fn_extract_basin(seglist, 'JA5_7480_0001')

```

**Comment by rburghol** [comment]
@jdkleiner I have done some more dev here.  Bringing in the points with existing tools, that can soon be used to do the bubble sizes as you mentioned above.  The only rub is that in a HUC8 size watershed things start to get kinda busy, though perhaps they wouldn't look so bad if I could figure out why my color scheme is lacking your earthtones.  The only real obstacle I see is that the labels overlap, which is a bit of a pain:

![image](https://user-images.githubusercontent.com/4571170/236533106-6f179a21-8e2c-42d1-8089-d3248e684ad1.png)

**Comment by rburghol** [comment]
Now, I have gone an alternate route for the container, and lost the busy stuff, as I used a variation on `plot_nhdplus()` that uses `comids` as input (instead of `bbox`) for the set of `flowlines`.  IT is less busy, but I can't get the catchment boundaries to show up crisply.

![image](https://user-images.githubusercontent.com/4571170/236549858-9c51f87b-5f6c-48fc-8c6c-2d66d92e19bd.png)

**Comment by rburghol** [comment]
Rivanna version:
![image](https://user-images.githubusercontent.com/4571170/236857298-81354499-ea9e-40f0-a178-b2cf5e8bba86.png)

**Comment by jdkleiner** [comment]
@rburghol This is great, glad to see you pushing this forward. Taking a look at some of this now, is your latest code housed here?: https://github.com/HARPgroup/vahydro/blob/master/R/mapping/EasyMapGen.R

Also, running into a snag, where is the function `fn_extract_basin()` defined?

**Comment by rburghol** [comment]
@jdkleiner yup that's the one (it's linked at the top of the issue below your original to trace the source).

The `fn_extract_basin()` is in the cia_utils.R set in hydrotools, you may need to update hydrotools? It was developed in 2021 but the linkage may not work in the module, so maybe you need to include cia_utils.R source link -- https://github.com/HARPgroup/hydro-tools/blob/d6ee09041c835aece8a338587c3473f4c28c6b2c/R/cia_utils.R#L174

Thanks for taking a look!!!

**Comment by rburghol** [comment]
@jdkleiner If you can verify your original version gets all those colors, my first B&W troubles happened with that one which makes me wonder if there is some config that I may lack?

**Comment by jdkleiner** [comment]
Thanks, looked like cia_utils.R got updated recently, needed to do a pull.

Just did a test of my earlier map version and your new version, both have color in the terrain layer. Maybe you need to update one of the spatial packages? (ggmap, ggsn, ggspatial)

![fig location_map_rb](https://user-images.githubusercontent.com/29379385/236933659-ab26619d-f251-49a5-851c-52df09aa3961.png)
- If we go the route of map extents based on watershed, we'll have to think of some smart ways to scale/zoom

**Comment by rburghol** [comment]
@jdkleiner 
> If we go the route of map extents based on watershed, we'll have to think of some smart ways to scale/zoom
Doesn't setting the extents eliminate the need to think about scale and zoom? The extent becomes the scale... no?

**Comment by rburghol** [comment]
Still no colors.  I did a `remove.packages()` for ggmap, ggspatial and ggsn then reinstalled.  Running R-4.2.2 -- and these color schemes have been with us for a lonnggg time.  tried the option for `get_map(...color = c('color')` to no avail...
